### PR TITLE
Preserve arg set status when cloning a kernel

### PR DESCRIPTION
**Pull request by callumfare**
This fixes test_api/clone_kernel from the CTS

This contribution is being made by Codeplay on behalf of Samsung.

### CODE REVIEW

**Review by kpet — CHANGES_REQUESTED** [review]
I don't think that's correct: not all arguments are necessarily set at the time where cloning happens. Could you add a test, here or, even better, in the CTS, for the following scenario:
- Create kernel object for a given kernel in a program
- Set only part of its arguments
- Attempt to enqueue it, check that `CL_INVALID_KERNEL_ARGS` is returned
- Clone the kernel
- Attempt to enqueue the cloned kernel, check that `CL_INVALID_KERNEL_ARGS` is returned

As to how to implement it, I think we should probably move `cvk_kernel::m_args_set` into the `cvk_kernel_argument_values` object and let the latter's copy constructor handle the copy of the exact state.

**Comment by callumfare** [comment]
Thanks for the review, that's what happens when I try to fix something last thing on a Friday! I've updated this PR to implement it as you suggested, and I've opened a new PR for the CTS to add an additional test case for `CL_INVALID_KERNEL_ARGS` - https://github.com/KhronosGroup/OpenCL-CTS/pull/1563

File: src/kernel.cpp [hunk 1]
```diff
@@ -134,6 +126,12 @@ cl_int cvk_kernel::set_arg(cl_uint index, size_t size, const void* value) {
     return ret;
 }
 
+bool cvk_kernel::args_valid() const {
+    return std::all_of(m_argument_values->args_set().cbegin(),
+                       m_argument_values->args_set().cend(),
+                       [](bool b) { return b; });
```
Comment (kpet) [inline src/kernel.cpp:132]: You could delegate this to the arguments object to reduce what it exposes to just an `args_valid` function.

**Review by kpet — APPROVED** [review]

**Comment by kpet** [comment]
Thanks! The change looks good. I've left a minor comment in case you're up for some more polishing but I'm fine merging the change as is and will do in a day or so if you don't push an update.

File: src/kernel.cpp [hunk 2]
```diff
@@ -134,6 +126,8 @@ cl_int cvk_kernel::set_arg(cl_uint index, size_t size, const void* value) {
     return ret;
 }
 
+bool cvk_kernel::args_valid() const { return m_argument_values->args_valid(); }
```
Comment (kpet) [inline src/kernel.cpp:129]: That could have been in the header but we've iterated enough on that change now :).

**Review by kpet — APPROVED** [review]
Thanks!

**Pull request merged**